### PR TITLE
Release videogames_back V1.0

### PR DESCRIPTION
**Pull request by AlejandraKeber**
## Hello code reviewer, 🌤️ 
Our group has 5 members: @sanieni6, @dicodiaz, @DavidEspinoG, @AlejandraKeber, and @NicolasAndrehh

Our Final Capstone Project is based on Videogames Reservations, an app that allows users to book a Video game. Users can see a list of available video games and can book a video game for a specific date range. Users can also see their reservations.  Admin users can add and delete video games.

This project was implemented in two repositories, one for the front-end with React, and one for the back-end with Ruby on Rails.
This PR is for the backend repository, which contains the API endpoints to be consumed in the front-end application.

## These are the links to the front-end: 
[Front-end repository](https://github.com/DavidEspinoG/mc_videogames_front)
[Front-end pull request](https://github.com/DavidEspinoG/mc_videogames_front/pull/43)

## 🛠️ In this back-end project we did the following changes:
- Create the Rails application, set up the project, and create the Kanban board.
- Design the ER diagram, and create the Postgresql database, model migrations, and populate the tables.
- Install devise and jwt-devise for authentication.
- Install cancancan and create user capabilities based on roles for authorization.
- Define model associations and validations.
- Create the controllers and API endpoints.
- Install rswag and create API documentation.
- Create Readme and License according to the Microverse template.
- Add unit and integration tests for models and controllers
- Refactor code to make it DRY

### Have a great day! 👋

### CODE REVIEW

File: app/controllers/reservations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  before_action :authenticate_user!
+
+  def index
+    @reservations = current_user.reservations.includes(:videogame).order(id: :desc)
+    render json: @reservations, include: :videogame
+  end
+
+  def new
+    @reservation = Reservation.new
+  end
```
Comment (mcrd25) [inline app/controllers/reservations_controller.rb]: Since it's an API, is this route necessary?

File: app/controllers/reservations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  def reservation_params
+    params.require(:reservation).permit(:videogame_id, :days, :total_price)
+  end
+
+  def user_from_token
+    jwt_payload = JWT.decode(request.headers['Authorization'].split[1],
+                             Rails.application.credentials.devise[:jwt_secret_key]).first
+    user_id = jwt_payload['sub']
+    User.find(user_id.to_s)
```
Comment (mcrd25) [inline app/controllers/reservations_controller.rb]: Missing exception Handling here, what if user_id does not exist?

File: app/controllers/reservations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  def double_check_user_verification
+    return false if request.headers['Authorization'].nil?
+
+    user = user_from_token
+    return true if user_signed_in? && current_user.id == user.id
+
+    false
```
Comment (mcrd25) [inline app/controllers/reservations_controller.rb]: Why is this necessary? Isn't ot a bit redundant?

File: app/controllers/videogames_controller.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
+class VideogamesController < ApplicationController
+  before_action :authenticate_user!, except: %i[index]
+
+  def index
+    @videogames = Videogame.all.order(id: :desc)
+    render json: @videogames
```
Comment (mcrd25) [inline app/controllers/videogames_controller.rb:6]: Consider Error Handling here as well (as seen in `create` and `destroy`)

File: app/controllers/videogames_controller.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  end
+
+  def show
+    @videogame = Videogame.find(params[:id])
+    render json: @videogame
```
Comment (mcrd25) [inline app/controllers/videogames_controller.rb:11]: Consider Error Handling here as well (as seen in `create` and `destroy`)

File: app/controllers/videogames_controller.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+  def videogame_params
+    params.require(:videogame).permit(:name, :description, :price_per_day, :photo)
+  end
+
+  def user_from_token
+    jwt_payload = JWT.decode(request.headers['Authorization'].split[1],
+                             Rails.application.credentials.devise[:jwt_secret_key]).first
+    user_id = jwt_payload['sub']
+    User.find(user_id.to_s)
```
Comment (mcrd25) [inline app/controllers/videogames_controller.rb]: See my comment on reservations

**Review by mcrd25 — CHANGES_REQUESTED** [review]
Hi Team,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Functional requirements are generally met 👍 
- routes are set up well 👍 
- You have tests 👍 
- API docs present 👍 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [OPTIONAL] Should always have positive and negative tests to ensure it works the way you expect it. Eg. What do you expect the output to be when it's not valid or when valid etc.
Consider making all your tests more extensive, not just this one. For this simple application, it may not catch a possible crippling bug, but in bigger applications, it definitely can. Consider this for all projects moving forward.


**Hope all is well, and that my comments were helpful. 😸**

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/videogames_controller.rb (hunk 4 above)
Comment (sanieni6) [reply app/controllers/videogames_controller.rb:6]: Hey there, the index endpoint is a public view so it's not required neither authentication nor authorization

File: app/controllers/videogames_controller.rb (hunk 5 above)
Comment (sanieni6) [reply app/controllers/videogames_controller.rb:11]: Hey there, the show endpoint is a public view so it's not required neither authentication nor authorization

File: README.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+```
+
+### Install
+
+Install this project with:
+
+ - `bundle install`
+ - `rails db:setup`
+
+### Usage
```
Comment (Meltrust) [inline README.md]: - [ ] When I try to create the database according to your instructions, this happens:

image

If this is a bug, kindly fix it so your database setup runs as intended.  If this is not a bug, this means your app requires special procedures to be run correctly.  If this is the case, kindly add specific instructions to your readme about how to setup and run the app properly.  

That way, all your users will be able to run your app smoothly. 👍 

_tip:  Is the app dependent on a .env file?  Are any special commands to be run before setting up the database?  Any extra setup?  If any of these is the case, add the proper instructions to your readme._

_note:  This means I couldn't create the database to check the API functionality and the tests locally.  These will be taken a look on the next review._

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi @AlejandraKeber,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Nice code organization ✔️ 
-  Cloud build is working well✔️ 





#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------
